server/middleware/applicationAuthMiddleware.test.ts

```diff
@@ -47,4 +47,36 @@ describe('applicationAuthMiddleware', () => {
     expect(response.status).toHaveBeenCalledWith(401)
     expect(response.render).toHaveBeenCalledWith('roleError')
   })
+
+  it('returns the handler when there are allowedPermissions specified and the user has one of those permissions', async () => {
+    const user = userDetailsFactory.build({ permissions: ['cas1_booking_create'] })
+    const response = createMock<Response>({ locals: { user } })
+
+    const auditedhandler = applicationAuthMiddleware(handler, {
+      allowedRoles: ['manager', 'matcher'],
+      allowedPermissions: ['cas1_booking_withdraw', 'cas1_process_an_appeal', 'cas1_booking_create'],
+    })
+
+    await auditedhandler(request, response, next)
+
+    expect(handler).toHaveBeenCalledWith(request, response, next)
+  })
+
+  it('redirects with an error when the user does not have the correct permission', async () => {
+    const user = userDetailsFactory.build({ roles: ['role_admin'], permissions: ['cas1_booking_create'] })
+    const response = createMock<Response>({ locals: { user } })
+
+    const loggerSpy = jest.spyOn(logger, 'error')
+
+    const auditedhandler = applicationAuthMiddleware(handler, {
+      allowedRoles: [],
+      allowedPermissions: ['cas1_booking_withdraw', 'cas1_process_an_appeal'],
+    })
+
+    await auditedhandler(request, response, next)
+
+    expect(loggerSpy).toHaveBeenCalledWith('User is not authorised to access this')
+    expect(response.status).toHaveBeenCalledWith(401)
+    expect(response.render).toHaveBeenCalledWith('roleError')
+  })
```

Comment:
Do you think it's worth adding a test that calls `applicationAuthMiddleware` with just the `handler`? 
The second parameter is optional, it might be worth demonstrating that the function can handle the `allowedPermissions` and `allowedRoles` being `undefined`.